[MNT] Deprecation message for `CyclicBoosting` changes

setoguchi-naoki opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://skpro.readthedocs.io/en/latest/contribute.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#232

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
- Add deprication warning
- Docstring
#### Does your contribution introduce a new dependency? If yes, which one?
No
<!--
If your contribution does add a new core dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core skpro package to a minimum.
-->
No
#### What should a reviewer concentrate their feedback on?
I'm not familiar with the detail for deprecation process and which future version we should set for removing default parameters.
<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?
No
<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).


<!--
Thanks for contributing!
-->

fkiraly commented on skpro/distributions/qpd.py:
```diff
@@ -160,6 +178,13 @@ def mean(self, lower=0.0, upper=np.inf):
+        comment = "Parameter 'lower' and 'upper' will be moved to \
+                   class parameters in version X.X.X. \
+                   To retain prior behaviour, set 'lower' to 0.0 \
+                   and 'upper' to large number explicitly"
```
should this not be `None` instead of a large number?

setoguchi-naoki replied:
It should be a large number here because 'upper' is used as the range of the differential. However, I made a mistake in the docstring for the default value of 'upper ', it is not None but 1e3 in the future version. In this case too, the large number should be set to the 'upper' parameter to cover the range of the user's dataset.

fkiraly reviewed (changes requested):
I think this is good, given the changes.

I would improve it as follows:

* the message in `__init__` needs only be raised if the changing parameters are invoked with a non-default, since that is the only case where behaviour changes or can break.

The parameters in the methods imo can be removed without deprecation message, as they were not API compliant

setoguchi-naoki commented:
> the message in __init__ needs only be raised if the changing parameters are invoked with a non-default, since that is the only case where behaviour changes or can break.

Thanks, I modified as following your comment. the default values of 'lower' and 'upper' will be removed in the future version. I made a wrong deprecation message. It is also modified. If the default value is removed, a deprecation message is always raised, as this can break the user's code.

> The parameters in the methods imo can be removed without deprecation message, as they were not API compliant

Sorry, I'm not sure I understand your comment. Which method did you mention?

setoguchi-naoki commented:
In addition, what future version should we specify in deprecation message?

fkiraly commented:
2.4.0 - the second next MINOR release.

I've opened a new PR here that covers only the regressor class: https://github.com/sktime/skpro/pull/329
The distribution should be covered by this: #254

fkiraly commented:
If we want to make further changes, these will have to wait until 2.5.0, as we will now proceed to the 2.3.0 release shortly.

The changes in #329 or #327 should not prevent us from making further changes to the distribution interface.

setoguchi-naoki commented:
Thank you for your contribution. As there are no tasks left for me to do, I have no problem closing this PR if it is unnecessary.

fkiraly commented:
Well, I dealt only with the `CyclicBoosting` class for deprecation.

We may stlil have to decide what to do with the distributions, in deprecation. For instance:

* all QPD_X distributions have parameters `lower` and `upper`. Are these needed? Or should we deprecate them? I think @felixwick mentioned this in his review of #327, and I replied that I left these because they were in the initial implementation.
* what about `dist_shape`?
* do we want to allow arbitrary distributions, incl `scipy` and `skpro` distributions as base distributions in QPD?

My advice would hence be, let's be clear about the target interfaces, from that it will become clear whether we need any further change/deprecation management, or not.

setoguchi-naoki commented:
> all QPD_X distributions have parameters lower and upper. Are these needed? Or should we deprecate them? I think @FelixWick mentioned this in his review of https://github.com/sktime/skpro/pull/327, and I replied that I left these because they were in the initial implementation.

As mentioned in #327 , the lower and upper parameters are not necessarily needed in all QPD classes if findiff is not used. Therefore, the upper parameter can be removed in QPD_S and the lower and upper parameters in QPD_U.

> what about dist_shape?

The dist_shape parameter is an option for fine-tuning the QPD_U. Therefore, it can be removed as a deprecated parameter if simplification of the QPD class interface is a priority. How do you think?

> do we want to allow arbitrary distributions, incl scipy and skpro distributions as base distributions in QPD?

It would be good to be able to specify scipy and skpro distributions, as you have implemented in #327. In this case, I don't think some support is needed in the deprecated process.

Let me check the situation. Should I add a deplication message to this PR to meet the #327 specification? Is there anything else I should consider?

FelixWick commented:
The shape parameter is only used for the extended QPD versions (Which include QPD_U. That means only QPD_S and QPD_B are "original" ones.), which I suggest to remove here.

fkiraly commented:
> As mentioned in https://github.com/sktime/skpro/pull/327 , the lower and upper parameters are not necessarily needed in all QPD classes if findiff is not used. Therefore, the upper parameter can be removed in QPD_S and the lower and upper parameters in QPD_U.

Ok, that makes sense, I was wondering what these were used for, already.

The need to use `findiff` is completely removed now, since the current native implementation computes `pdf` explicitly (just use the chain rule 😁 ). So let's deprecate the unused parameters then.

> The dist_shape parameter is an option for fine-tuning the QPD_U. Therefore, it can be removed as a deprecated parameter if simplification of the QPD class interface is a priority. How do you think?

I agree it should be removed. I would actually replace it by:

* renaming `version` to a more appropriate name, e.g., `base_dist`
* allowing `base_dist` to be a parametric `skpro` distirbution, which can have its own parameters. If users want to pass `dist_shape`, they can pass it as a parameter to `base_dist`.

fkiraly commented:
> Let me check the situation. Should I add a deplication message to this PR to meet the https://github.com/sktime/skpro/pull/327 specification? Is there anything else I should consider?

This might require extending the logic, at least addition of a test case that passes an `skpro` distirbution.

I would also strongly suggest to rename the parameter `version`, as it has strong connotations with versioning, e.g., semver.

setoguchi-naoki commented:
@fkiraly 
I modfied deplication message like

- Removal of uppers and dist_shape from some QPD classes
- Rename version to base_dist

> at least addition of a test case that passes an skpro distirbution.

The `skpro` distribution has not yet been introduced and tested in qpd.py. Should it be done in other branches?

fkiraly commented:
Great!

I think you're quite out of sync with your branch since the new release - can you merge `main` into your branch, and ensure to resolve all conflicts? Only the files related to `CyclicBootsing` should be changed.

setoguchi-naoki commented:
I think I have resolved the conflicts in the main branch.

fkiraly commented:
> I think I have resolved the conflicts in the main branch.

👍 yes, looks like it!

fkiraly reviewed (changes requested):
Great!

One minor thing, we should ensure that users can already move to the future behaviour, i.e., have a chance to update their code in advance. That would mean adding the `base_dist` parameter already right now, an dif it is passed, it overrides `version`.

Have a look here how renaming a parameter works, including a suggested warning message: https://www.sktime.net/en/stable/developer_guide/deprecation.html#id2

setoguchi-naoki commented:
Thank you. Fixes for parameter renaming have been applied.

fkiraly reviewed (approved):
Made some minor tweaks around the docstrings (moved to the "final" form), and release manager actions.
